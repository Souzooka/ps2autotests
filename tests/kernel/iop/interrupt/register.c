#include <stdio.h>
#include <intrman.h>
#include <thsemap.h>

static int g_handlerResult = -1;
static int g_handlerSema = 0;

void intrHandler(int value) {
	g_handlerResult = value;
	iSignalSema(g_handlerSema);
}

void createHandlerSema() {
	iop_sema_t semaInfo = {};
	semaInfo.attr    = 0;
	semaInfo.option  = 0;
	semaInfo.initial = 0;
	semaInfo.max     = 1;
	g_handlerSema = CreateSema(&semaInfo);
}

void testMultipleRegister() {
	int oldStat = 0, oldIntr = 0;

	createHandlerSema();

	printf("multiple register:\n");
	CpuSuspendIntr(&oldStat);
	{
		//Make sure no handler is currently registered
		ReleaseIntrHandler(IOP_IRQ_VBLANK);
		DisableIntr(IOP_IRQ_VBLANK, &oldIntr);

		int firstRegisterResult = RegisterIntrHandler(IOP_IRQ_VBLANK, 0, (void *)&intrHandler, (void *)0xAAAA);
		printf("\tregister when no handler is present: %d\n", firstRegisterResult);
		
		int secondRegisterResult = RegisterIntrHandler(IOP_IRQ_VBLANK, 0, (void *)&intrHandler, (void *)0xBBBB);
		printf("\tregister when handler is already present: %d\n", secondRegisterResult);
		
		EnableIntr(IOP_IRQ_VBLANK);
	}
	CpuResumeIntr(oldStat);
	
	//Wait for handler to be called at least once
	WaitSema(g_handlerSema);
	
	CpuSuspendIntr(&oldStat);
	{
		printf("\thandler result: %d\n", g_handlerResult);

		ReleaseIntrHandler(IOP_IRQ_VBLANK);
		DisableIntr(IOP_IRQ_VBLANK, &oldIntr);
	}
	CpuResumeIntr(oldStat);
	
	DeleteSema(g_handlerSema);
}

void testInvalidParams() {
	int oldStat = 0;

	printf("invalid params:\n");

	CpuSuspendIntr(&oldStat);
	{
		int invalidIntrLineResult = RegisterIntrHandler(~0, 0, (void *)&intrHandler, NULL);
		printf("\tinvalid interrupt line: %d\n", invalidIntrLineResult);
		
		int invalidModeResult = RegisterIntrHandler(IOP_IRQ_VBLANK, ~0, (void *)&intrHandler, NULL);
		ReleaseIntrHandler(IOP_IRQ_VBLANK);
		printf("\tinvalid mode line: %d\n", invalidModeResult);

		int nullHandlerResult = RegisterIntrHandler(IOP_IRQ_VBLANK, 0, NULL, NULL);
		ReleaseIntrHandler(IOP_IRQ_VBLANK);
		printf("\tnull handler line: %d\n", nullHandlerResult);
	}
	CpuResumeIntr(oldStat);
}

int _start(int argc, char **argv) {
	printf("-- TEST BEGIN\n");
	
	testMultipleRegister();
	testInvalidParams();

	printf("-- TEST END\n");
	
	return 0;
}
