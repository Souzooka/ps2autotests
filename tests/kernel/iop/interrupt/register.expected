-- TEST BEGIN
multiple register:
  register when no handler is present: 0
  register when handler is already present: -104
  handler result: 43690
null handler:
  registering null handler: 0
  registering null handler after null handler was registered: 0
  registering null handler after non-null handler was registered: -104
  registering non-null handler after null handler was registered: 0
invalid params:
  invalid interrupt line: -101
  invalid mode (INT_MIN): 0
  invalid mode (INT_MAX): 0
-- TEST END
